urlc98c3b97d2_Trimax_venta_tgz-pJ8-com_gesoftware_venta_structures_map_MapTestJ8/src/com/gesoftware/venta/network/Server.java

```diff
@@ -0,0 +1,570 @@
+
+                /* That's it */
+                LoggingUtility.core("Sender thread closed for client: " + m_ID);
+            } /* End of 'ClientSender::run' method */
+        } /* End of 'ClientSender' class */
+
+        /* *
+         * METHOD: Disconnect client from server
+         * AUTHOR: Eliseev Dmitry
+         * */
+        private void disconnectClient() {
+            /* Remove client from internal data structure */
+            removeClient(m_ID);
+
+            /* Stop client thread */
+            m_IsClientActive = false;
+
+            /* Notify sender thread */
+            synchronized (m_SenderMonitor) {
+                m_SenderMonitor.notify();
+            }
+
+            /* Closing socket */
+            if (m_Socket != null)
+                synchronized (m_Socket) {
+                    try {
+                        m_Socket.close();
+                    } catch (final IOException ignored) {}
+                }
+        } /* End of 'Client::disconnectClient' method */
+
+        /* *
+         * METHOD: Client class constructor
+         *  PARAM: [IN] socket - connected client socket
+         * AUTHOR: Eliseev Dmitry
+         * */
+        public Client(final Socket socket) {
+            this.m_ID        = HashUtility.generateHash(socket.toString());
+            this.m_Socket    = socket;
+            this.m_LastQuery = new Timer();
+        } /* End of 'Client::Client' method */
+
+        /* *
+         * METHOD: Starts client processing thread
+         * AUTHOR: Eliseev Dmitry
+         * */
+        public final void start() {
+            /* Creation and starting sender thread */
+            m_Sender       = new ClientSender();
+            m_ThreadSender = new Thread(m_Sender);
+            m_ThreadSender.start();
+
+            /* Creation and starting receiver thread */
+            m_Receiver       = new ClientReceiver();
+            m_ThreadReceiver = new Thread(m_Receiver);
+            m_ThreadReceiver.start();
+        } /* End of 'Client::start' method */
+
+        /* *
+         * METHOD: Disconnects client immediately
+         * AUTHOR: Eliseev Dmitry
+         * */
+        public final void disconnect() {
+            m_Receiver.stop();
+        } /* End of 'Client::disconnect' method */
+
+        /* *
+         * METHOD: Sends a message to client
+         * RETURN: True if success, false otherwise
+         *  PARAM: [IN] message - reference to message to send
+         * AUTHOR: Eliseev Dmitry
+         * */
+        public final synchronized boolean send(final Message message) {
+            try {
+                m_Messages.put(message);
+
+                /* Inform sender thread about new messages */
+                synchronized (m_SenderMonitor) {
+                    m_SenderMonitor.notify();
+                }
+
+                /* That's it (send successfully) */
+                return true;
+            } catch (final InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)